confidant/services/certificate_authority/customca.py

```diff
@@ -0,0 +1,225 @@
+    "crt": {"type": "string", "required": True},
+    "key": {"type": "string", "required": True},
+    "passphrase": {"type": "string", "required": True, "nullable": True},
+    "kid": {"type": "string", "required": True},
+}
+
+
+class CustomCertificateAuthority(CertificateAuthorityBase):
+    """Custom Certificate Authority
+
+    Args:
+        CertificateAuthorityBase (_type_): Base class for Certificate Authority
+    """
+
+    def __init__(self, ca_env: str):
+        self.ca_id = ca_env
+        self.active_ca_id = None
+        self.ca_json = self._get_ca_in_json(ca_env)
+        self.ca_certificate = self._load_ca_certificate(self.ca_json)
+        self.root_ca_certificate = self._load_rootca_certificate(self.ca_json)
+        self.ca_private_key = self._load_private_key(self.ca_json)
+        self.ca_chain = self._load_ca_chain()
+        self.settings = settings.CUSTOM_CA_SETTINGS
+
+    def _get_ca_in_json(self, ca_env: str):
+        if (
+            not CUSTOM_CERTIFICATE_AUTHORITIES
+            or ca_env not in CUSTOM_CERTIFICATE_AUTHORITIES
+        ):
+            logger.error("Custom CA %s not found", ca_env)
+            raise CertificateAuthorityNotFoundError(
+                f"Custom CA {ca_env} not found"
+            )
+        if not CUSTOM_CA_ACTIVE_KEYS or ca_env not in CUSTOM_CA_ACTIVE_KEYS:
+            logger.error("Custom CA %s has no active keys", ca_env)
+            raise CertificateAuthorityNotFoundError(
+                f"Custom CA {ca_env} has no active keys"
+            )
+        validator = Validator(CUSTOM_CA_SCHEMA)
+        active_ca_id = CUSTOM_CA_ACTIVE_KEYS[ca_env]
+        self.active_ca_id = active_ca_id
+        active_ca = [
+            ca
+            for ca in CUSTOM_CERTIFICATE_AUTHORITIES[ca_env]
+            if validator.validate(ca) and ca["kid"] == active_ca_id
+        ]
+        if not active_ca:
+            logger.error("Custom CA %s has no active keys", ca_env)
+            raise CertificateAuthorityNotFoundError(
+                (
+                    f"Custom CA {ca_env} has no matching valid active keys for "
+                    f"{active_ca_id}"
+                )
+            )
+        return active_ca[0]
+
+    def _load_ca_certificate(self, ca_json):
+        return x509.load_pem_x509_certificate(ca_json["crt"].encode("utf-8"))
+
+    def _load_rootca_certificate(self, ca_json):
+        if "rootcrt" not in ca_json or not ca_json["rootcrt"]:
+            logger.warning("Custom CA has no root CA certificate provided")
+            return None
+        return x509.load_pem_x509_certificate(
+            ca_json["rootcrt"].encode("utf-8")
+        )
+
+    def _load_ca_chain(self):
+        # Get the certificate in PEM format
+        intermediate_ca_pem = self.encode_certificate(self.ca_certificate)
+        root_ca_pem = self.encode_certificate(self.root_ca_certificate)
+        return intermediate_ca_pem + root_ca_pem
```

Comment:
self.root_ca_certificate can be None? if so intermediate_ca_pem + root_ca_pem will throw TypeError